#### 3) a.

In [ ]:
import pandas as pd

df = pd.read_csv('./data/300110-11-accidentes-bicicleta-csv.csv', sep=';', encoding='utf-8')

display(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   num_expediente        891 non-null    str    
 1   fecha                 891 non-null    str    
 2   hora                  891 non-null    str    
 3   localizacion          891 non-null    str    
 4   numero                891 non-null    str    
 5   cod_distrito          891 non-null    int64  
 6   distrito              891 non-null    str    
 7   tipo_accidente        891 non-null    str    
 8   estado_meteorológico  865 non-null    str    
 9   tipo_vehículo         891 non-null    str    
 10  tipo_persona          891 non-null    str    
 11  rango_edad            891 non-null    str    
 12  sexo                  891 non-null    str    
 13  cod_lesividad         795 non-null    float64
 14  tipo_lesividad        795 non-null    str    
 15  coordenada_x_utm      891 non-null

None

In [7]:
nominal_threshold = 10
for col in df.columns:
    if df[col].dtype == "str":
        print(f"{col:>20} -> Categórica Nominal")
        continue
    else:
        is_discrete = True

        for value in df[col]:
            if value % 1 > 0:
                is_discrete = False
                break

    if is_discrete:
        uniques = df[col].nunique()
        print(f"{col:>20} -> Numérica Discreta / Categórica {'Ordinal' if uniques > nominal_threshold else 'Nominal'}")
    else:
        print(f"{col:>20} -> Numérica Contínua")

      num_expediente -> Categórica Nominal
               fecha -> Categórica Nominal
                hora -> Categórica Nominal
        localizacion -> Categórica Nominal
              numero -> Categórica Nominal
        cod_distrito -> Numérica Discreta / Categórica Ordinal
            distrito -> Categórica Nominal
      tipo_accidente -> Categórica Nominal
estado_meteorológico -> Categórica Nominal
       tipo_vehículo -> Categórica Nominal
        tipo_persona -> Categórica Nominal
          rango_edad -> Categórica Nominal
                sexo -> Categórica Nominal
       cod_lesividad -> Numérica Discreta / Categórica Nominal
      tipo_lesividad -> Categórica Nominal
    coordenada_x_utm -> Categórica Nominal
    coordenada_y_utm -> Categórica Nominal
    positiva_alcohol -> Categórica Nominal
      positiva_droga -> Numérica Discreta / Categórica Nominal


3) a. **Conclusões**:
* Pelos mesmos motivos apresentados nos exercícios anteriores, seguirei a mesma abordagem para identificar variáveis numéricas e categóricas, especificando se são contínuas, discretas, ordinais ou nominais.
* Apesar de conseguirmos identificar que por exemplo que as variáveis que representam coordenadas (*coordenada_x_utm* e *coordenada_y_utm*) e timestamps (*fecha* e *hora*) poderiam ser representadas num espaço contínuo, mas para isso elas teriam que ser transformadas.

| variável | tipo |
|-|-|
|       num_expediente | Categórica Nominal |
|                fecha | Categórica Nominal |
|                 hora | Categórica Nominal | 
|         localizacion | Categórica Nominal |
|               numero | Categórica Nominal |
|         cod_distrito | Numérica Discreta / Categórica Ordinal |
|             distrito | Categórica Nominal |
|       tipo_accidente | Categórica Nominal |
| estado_meteorológico | Categórica Nominal |
|        tipo_vehículo | Categórica Nominal |
|         tipo_persona | Categórica Nominal |
|           rango_edad | Categórica Nominal |
|                 sexo | Categórica Nominal |
|        cod_lesividad | Numérica Discreta / Categórica Nominal |
|       tipo_lesividad | Categórica Nominal |
|     coordenada_x_utm | Categórica Nominal |
|     coordenada_y_utm | Categórica Nominal |
|     positiva_alcohol | Categórica Nominal |
|       positiva_droga | Numérica Discreta / Categórica Nominal |

#### 3) b.

In [8]:
missing_values = df.isnull().sum()
print(missing_values[missing_values > 0])

estado_meteorológico     26
cod_lesividad            96
tipo_lesividad           96
positiva_alcohol          2
positiva_droga          887
dtype: int64


3) b. **Conclusões**:
* Variáveis com dados em falta:
    * estado_meteorológico     26
    * cod_lesividad            96
    * tipo_lesividad           96
    * positiva_alcohol          2
    * positiva_droga          887

#### 3) c.

In [10]:
count_uni_lesividad = df['cod_lesividad'].nunique()
uni_lesividad = df['cod_lesividad'].unique()

print(f"A variável 'cod_lesividad' assume {count_uni_lesividad} valores diferentes.")
print(f"Os valores são: {uni_lesividad}")
print(f"A variável é do tipo: {df['cod_lesividad'].dtype}")

A variável 'cod_lesividad' assume 8 valores diferentes.
Os valores são: [ 3.  2.  7. 14.  1. nan  6.  5.  4.]
A variável é do tipo: float64


3) c. **Conclusões**:
* A variável 'cod_lesividad' assume 8 valores diferentes.
* Os valores são: [ 3.  2.  7. 14.  1. nan  6.  5.  4.]
* A variável é do tipo: *float64*, apesar de nenhum dos valores tem parte decimal, o que induz que poderia ser convertido para *int64* sem perda de informação

#### 3) d.

Dados recolhidos do pdf:

* 01 Atención en urgencias sin posterior ingreso. - LEVE
* 02 Ingreso inferior o igual a 24 horas - LEVE
* 03 Ingreso superior a 24 horas. - GRAVE
* 04 Fallecido 24 horas - FALLECIDO
* 05 Asistencia sanitaria ambulatoria con posterioridad - LEVE
* 06 Asistencia sanitaria inmediata en centro de salud o mutua - LEVE
* 07 Asistencia sanitaria sólo en el lugar del accidente - LEVE
* 14 Sin asistencia sanitaria
* 77 Se desconoce
* En blanco Sin asistencia sanitaria

Numa primeira análise, a variável parece ser Categórica Nominal, pois surge sob a forma de números (códigos) sem um significado de ordem imediato e sem contexto. No entanto podemos ver que a Câmara Municipal de Madrid já agrupa estes códigos em categorias mais amplas e hierárquicas (LEVE, GRAVE, FALLECIDO), que representam uma progressão na gravidade do desfecho do acidente, desde sair ileso (sem assistência) até ao pior cenário (falecido). Assim, poderia ser realizada uma conversão numa variável Categórica Ordeal.

Vantagens desta abordagem para o seu modelo:
* Reduz o ruído: O modelo não tem de adivinhar a diferença matemática entre "urgência sem internamento" e "assistência ambulatorial", sabendo apenas que ambos são casos "Leves".
* Mantém a ordem matemática: Um valor 3 (Falecido) pesa muito mais do que um 0 (Sem assistência), o que ajuda algoritmos baseados em árvores de decisão (como Random Forest) ou regressões a encontrar padrões mais lógicos. 

Um exemplo de codificação par este processso de conversão seria como descrito no bloco abaixo:

In [12]:
lesividad_map = {
    14: 0,  # Sin asistencia sanitaria
    1: 1,  # LEVE: Atención en urgencias sin posterior ingreso
    2: 1,  # LEVE: Ingreso inferior o igual a 24 horas
    5: 1,  # LEVE: Asistencia sanitaria ambulatoria con posterioridad
    6: 1,  # LEVE: Asistencia sanitaria inmediata en centro de salud
    7: 1,  # LEVE: Asistencia sanitaria sólo en el lugar del accidente
    3: 2,  # GRAVE: Ingreso superior a 24 horas
    4: 3,  # FALLECIDO: Fallecido 24 horas
    77: 0  # SE DESCONOCE: Definimos como 0, pois não temos informação sobre a lesividade
}
# Nans (En blanco) serão tratados como 0 (sem lesividade)
df['lesividad_ordinal'] = df['cod_lesividad'].map(lambda x: lesividad_map.get(x) if x in lesividad_map else 0)
print(df[['cod_lesividad', 'lesividad_ordinal']].head())

   cod_lesividad  lesividad_ordinal
0            3.0                  2
1            2.0                  1
2            7.0                  1
3           14.0                  0
4            7.0                  1


#### 3) e.

In [18]:
alcohol_proportions = df['positiva_alcohol'].dropna().value_counts(normalize=True)
print(alcohol_proportions)

positive_pct = alcohol_proportions['S'] * 100
print(f"\nProporção de testes de álcool positivo: {positive_pct:.2f}%")

positiva_alcohol
N    0.987627
S    0.012373
Name: proportion, dtype: float64

Proporção de testes de álcool positivo: 1.24%


Proporção de testes de álcool positivo: 1.24%